src/qmplay2/YouTubeDL.cpp

```diff
@@ -308,26 +315,38 @@ QStringList YouTubeDL::exec(const QString &url, const QStringList &args, QString
             return {};
         }
 
-        //[Title], url, JSON, [url, JSON]
-        for (int i = result.count() - 1; i >= 0; --i)
+        // construct a sorted map of the available heights and the corresponding stream URLs
+        QMap<int,QString> height2Url;
+        for (int i = 1 ; i < result.count(); ++i)
         {
             if (i > 0 && result.at(i).startsWith('{'))
             {
-                const QString url = result.at(i - 1);
-
                 const QJsonDocument json = QJsonDocument::fromJson(result.at(i).toUtf8());
                 for (const QJsonValue &formats : json.object()["formats"].toArray())
                 {
-                    if (url == formats.toObject()["url"].toString())
-                        QMPlay2Core.addCookies(url, formats.toObject()["http_headers"].toObject()["Cookie"].toString().toUtf8());
+                    const auto streamUrl = formats.toObject()["url"].toString();
+                    const int height = formats.toObject()["height"].toInt();
+                    QMPlay2Core.addCookies(streamUrl, formats.toObject()["http_headers"].toObject()["Cookie"].toString().toUtf8());
+                    // there can be multiple formats with the same height
+                    height2Url.insertMulti(height, streamUrl);
                 }
-
-                result.removeAt(i);
             }
         }
 
+        // construct the return list
+        QStringList retList(result.at(0));
+        auto urlList = height2Url.values().toStdList();
+        // reverse-sort the URL list so that the higher resolutions get priority
+        urlList.reverse();
+        for (const auto streamUrl : urlList)
```

Comment:
Maybe you can iterate over `height2Url` and prepend values instead of copying them to `QList<QString>` and then conversion to std linked list?